src/lib/serialization/schema.ts

```diff
@@ -0,0 +1,255 @@
+
+/**
+ * Describes the mapping from Model types to the corresponding JSON output type.
+ */
+export type ModelToObject<T> = T extends Array<infer U> ? _ModelToObject<U>[] : _ModelToObject<T>;
+
+// Order matters here. Some types are subtypes of other types.
+type _ModelToObject<T> =
+    // Reflections
+    T extends M.ReflectionGroup ? JSONOutput.ReflectionGroup :
+    T extends M.ReflectionCategory ? JSONOutput.ReflectionCategory :
+    T extends M.SignatureReflection ? JSONOutput.SignatureReflection :
+    T extends M.ParameterReflection ? JSONOutput.ParameterReflection :
+    T extends M.DeclarationReflection ? JSONOutput.DeclarationReflection | JSONOutput.ReflectionPointer :
+    T extends M.TypeParameterReflection ? JSONOutput.TypeParameterReflection :
+    T extends M.ProjectReflection ? JSONOutput.ProjectReflection :
+    T extends M.ContainerReflection ? JSONOutput.ContainerReflection :
+    T extends M.Reflection ? JSONOutput.Reflection :
+    // Types
+    T extends M.ArrayType ? JSONOutput.ArrayType :
+    T extends M.IntersectionType ? JSONOutput.IntersectionType :
+    T extends M.IntrinsicType ? JSONOutput.IntrinsicType :
+    T extends M.ReferenceType ? JSONOutput.ReferenceType :
+    T extends M.ReflectionType ? JSONOutput.ReflectionType :
+    T extends M.StringLiteralType ? JSONOutput.StringLiteralType :
+    T extends M.TupleType ? JSONOutput.TupleType :
+    T extends M.UnknownType ? JSONOutput.UnknownType :
+    T extends M.Type ? JSONOutput.SomeType : // Technically AbstractType, but the union is more useful
+    // Miscellaneous
+    T extends M.Comment ? JSONOutput.Comment :
+    T extends M.CommentTag ? JSONOutput.CommentTag :
+    T extends DecoratorWrapper ? JSONOutput.Decorator :
+    T extends SourceReferenceWrapper ? JSONOutput.SourceReference :
+    never;
+
+/**
+ * Contains interfaces which describe the JSON output. Each interface is related to a specific type of serializer.
+ *
+ * ## Plugins
+ * Plugins which modify the serialization process can use declaration merging
+ * to add custom properties to the exported interfaces.
+ * For example, if your custom serializer adds a property to all [[Reflection]] objects:
+ * ```ts
+ * declare module 'typedoc/dist/lib/serialization/schema' {
+ *     export namespace JSONOutput {
+ *         export interface AbstractReflection {
+ *             myCustomProp: boolean
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * If a plugin defines a new Model type, [[ModelToObject]] will not pick up the serializer type.
+ * To fix this, use declaration merging to augment the [[Serializer]] class.
+ * ```ts
+ * declare module 'typedoc/dist/lib/serialization/serializer' {
+ *     export interface Serializer {
+ *         toObject(value: CustomModel, obj?: Partial<CustomModel>): CustomOutput
+ *     }
+ * }
+ * ```
+ *
+ * For documentation on the JSON output properties, view the corresponding model.
+ */
+export namespace JSONOutput {
```

Comment:
I believe my idea was to avoid polluting the global scope with ~40 new types.

The new top level export makes sense, but I think the way to do it should probably be to not publish TypeDoc within the `dist/lib` folder, this would be a massive breaking change, since basically all plugins import the `Component` decorator from `typedoc/dist/lib/...`

Reply:
I've moved these out of the namespace, there was no need as avoiding the pollution could be done by re-exporting under a specific name later. I still think the `typedoc/serialization` module is a good idea, but again, breaking changes that don't need to happen yet & should be thought through carefully.